RIOT-OS/RIOT
# pkg/wolfssl: implement sock_dtls

### Contribution description

This PR implements the synchronous part DTLS sock API for wolfSSL. An async implementation will be done as a follow-up.

This PR also moves the existing sock_tls implementation provided by the wolfSSL team to its own directory under the contrib directory. sock_tls is a wrapper around wolfSSL functions to setup and destroys sessions when using wolfSSL directly. Although the similarity the name is similar, it is not related to the sock_dtls API.

### Testing procedure

Use the `dtls-sock` example to test. Make sure that in the Makefile, `USEPKG += wolfssl` and one of  `USEMODULE += wolfssl_psk` or `USEMODULE += wolfcrypt_ecc` are uncommented. Try sending messages between client and server. The behavior should be the same as when using tinydtls apart from the issues mentioned below.


#### Known Issues

I tried to make sure this implementation behaves as close as possible to the existing tinydtls implementation but there are some parts that are harder to do so:

- wolfSSL uses a different timeout system during a handshake compared to RIOT
  - wolfSSL tracks timeout using 3 struct member `dtls_timeout_init`, `dtls_timeout`, and `dtls_timeout_max`
  - During a handshake process, `dtls_timeout` is set to `dtls_timeout_init` at the start and multiplied by 2 each time the `_recv` callback returns until it reaches `dtls_timeout_max` where it will return
  - this means if the server is given a `timeout` value of 1 second to listen to new connections, it might actually wait for more than that
  - this PR works around this by setting all three timeout value that wolfSSL uses to the same value so that it will return directly after one call to the `_recv` callback
- there are no `SOCK_NO_TIMEOUT` equivalent in wolfSSL which means the server might return an error when waiting for a long time even though timeout is set to `SOCK_NO_TIMEOUT`
- wolfSSL will not send data if len is 0. The handshake will succeed and it will not report any error but Wireshark does not show any data transmitted
- wolfSSL will automatically call the `_recv` callback to listen to handshake packets after a new ClientHello is sent which is not the expected behavior of the `sock_dtls_session_init` function
    - This PR currently uses non-blocking options of wolfSSL to make it return directly after sending the first ClientHello but sometimes it might already receives other handshake packets before returning. 

### Useful Links

[Documentation of the wolfSSL API](https://www.wolfssl.com/doxygen/wolfssl_API.html)
